tee-worker/litentry/core/credentials/src/credentials.rs

```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+#[derive(Serialize, Deserialize, Encode, Decode, Clone, Debug, PartialEq, Eq, TypeInfo)]
+#[serde(rename_all = "camelCase")]
+pub struct Credential {
+	#[serde(rename = "@context")]
+	pub context: Vec<String>,
+	pub id: String,
+	pub subject: String,
+	#[serde(rename = "type")]
+	pub types: Vec<CredentialType>,
+	pub credential_subject: CredentialSubject,
+	pub issuer: Issuer,
+	pub issuance_date: Option<String>,
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub expiration_date: Option<String>,
+	pub proof: Proof,
+}
+
+impl Credential {
+	pub fn from_template(s: &str, who: &AccountId) -> Result<Self, Error> {
+		let mut vc: Self =
+			serde_json::from_str(s).map_err(|err| Error::ParseError(format!("{}", err)))?;
+		vc.subject = account_id_to_string(who);
+		vc.issuance_date = Some(now());
+		vc.validate_unsigned()?;
+		Ok(vc)
+	}
+
+	pub fn to_json(&self) -> Result<String, Error> {
+		let json_str =
+			serde_json::to_string(&self).map_err(|err| Error::ParseError(format!("{}", err)))?;
+		Ok(json_str)
+	}
+
+	pub fn validate_unsigned(&self) -> Result<(), Error> {
+		if !self.types.contains(&CredentialType::VerifiableCredential) {
+			return Err(Error::EmptyCredentialType)
+		}
+
+		if self.subject.is_empty() {
+			return Err(Error::EmptySubject)
+		}
+
+		if self.issuance_date.is_none() {
+			return Err(Error::EmptyIssuanceDate)
+		}
+
+		Ok(())
+	}
+
+	pub fn validate(&self) -> Result<(), Error> {
+		self.validate_unsigned()?;
+
+		if self.credential_subject.is_empty() {
+			return Err(Error::EmptyCredentialSubject)
+		}
+
+		// ToDo: validate issuer
+		// if self.issuer.is_empty() {
+		// 	return Err(Error::EmptyCredentialIssuer)
+		// }
+
+		//ToDo: validate the proof timestamp that is must be equal or after issuance timestamp
+		if self.proof.created.is_none() {
+			return Err(Error::InvalidDateOrTimeError)
+		}
+
+		// ToDo: validate proof
+		// if self.proof.is_empty() {
+		// 	return Err(Error::EmptyCredentialProof)
+		// }
+
+		//ToDo: validate proof signature
+
+		let exported = self.to_json()?;
+		if exported.len() > MAX_CREDENTIAL_SIZE {
+			return Err(Error::CredentialIsTooLong)
+		}
+
+		Ok(())
+	}
+
+	pub fn validate_schema(&self) -> Result<(), Error> {
+		//ToDo: fetch and check schema status from Parentchain storage
+		Ok(())
+	}
+
+	pub fn generate_unsigned_credential(
+		assertion: &Assertion,
+		who: &AccountId,
+	) -> Result<Credential, Error> {
+		debug!("generate unsigned credential {:?}", assertion);
+		match assertion {
+			Assertion::A1 => {
+				let raw = include_str!("templates/a1.json");
+				let credential: Credential = Credential::from_template(raw, who)?;
+				Ok(credential)
+			},
+			_ => Err(Error::UnsupportedAssertion),
+		}
+	}
+
+	pub fn add_assertion_a1(&mut self, _web2_cnt: i32, _web3_cnt: i32) {
+		self.credential_subject.assertions = "\"or\": [{\"src\": \"$web2_account_cnt\", \"op\": \">\", \"dsc\": \"0\",},{\"src\": \"$web3_account_cnt\", \"op\": \">\", \"dsc\": \"0\",}]".to_string();
```

Comment:
maybe let's use raw string literals for better readability, something like this:
```
format!(r#"{{"type": "type1", "type2": {}}}"#, var1);
```